File: 2023/cpp/senior/senior5.cpp
#include <bits/stdc++.h>
using namespace std;

unordered_set<int> numbers{};
const int N = pow(3, 19);

void fill(int a, int n, int k) {
    if (k >= N) {
        for (int i = (long) a * n / k; i <= (long) (a + 1) * n / k; i++) {
            numbers.insert(i);
        }
    } else {
        fill(a * 3, n, k * 3);
        fill(a * 3 + 2, n, k * 3);
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    fill(0, n, 1);

    vector<int> answers{};
    for (int i : numbers) {
        unordered_set<int> s;
        s.insert(0);
        s.insert(n);
        long a = i;
        bool covered = false;
        while (true) {
            if (3 * a <= n) {
                a *= 3;
                if (s.find(a) != s.end()) {
                    covered = false;
                    break;
                }
                s.insert(a);
            } else if (3 * a >= n * 2) {
                a = a * 3 - n * 2;
                if (s.find(a) != s.end()) {
                    covered = false;
                    break;
                }
                s.insert(a);
            } else {
                covered = true;
                break;
            }
        }
        if (!covered) {
            answers.emplace_back(i);
        }
    }

    sort(answers.begin(), answers.end());
    for (int i : answers) {
        cout << i << endl;
    }
}

File: 2023/java/senior/Senior5.java
import java.util.*;

public class Senior5 {
    private static Set<Integer> numbers = new HashSet<>();
    private static int N = (int) Math.pow(3, 19);

    private static void fill(int a, int n, int k) {
        if (k >= N) {
            for (int i = (int) ((long) a * n / k); i <= (long) (a + 1) * n / k; i++) {
                numbers.add(i);
            }
        } else {
            fill(a * 3, n, k * 3);
            fill(a * 3 + 2, n, k * 3);
        }
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        fill(0, n, 1);

        List<Integer> answers = new ArrayList<>();
        for (int i : numbers) {
            Set<Integer> s = new HashSet<>();
            s.add(0);
            s.add(n);
            long a = i;
            boolean covered = false;
            while (true) {
                if (3 * a <= n) {
                    a *= 3;
                    if (s.contains((int) a)) {
                        covered = false;
                        break;
                    }
                    s.add((int) a);
                } else if (3 * a >= n * 2) {
                    a = a * 3 - n * 2;
                    if (s.contains((int) a)) {
                        covered = false;
                        break;
                    }
                    s.add((int) a);
                } else {
                    covered = true;
                    break;
                }
            }
            if (!covered) {
                answers.add(i);
            }
        }

        answers.sort(Comparator.naturalOrder());
        for (int i : answers) {
            System.out.println(i);
        }
    }
}
